clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/BlobSnapshotManager.java

```diff
@@ -22,66 +25,142 @@
+      BlobTransferPartitionMetadata blobTransferPartitionMetadata) {
+    boolean snapshotStale = isSnapshotStale(topicName, partitionId);
+    if (snapshotStale) {
+      try {
+        createSnapshot(topicName, partitionId);
+        // update the snapshot timestamp to reflect the latest snapshot creation time
+        snapshotTimestamps.get(topicName).put(partitionId, System.currentTimeMillis());
+        // update the snapshot offset record to reflect the latest snapshot offset
+        snapshotMetadataRecords.get(topicName).put(partitionId, blobTransferPartitionMetadata);
+        LOGGER.info("Successfully created snapshot for topic {} partition {}", topicName, partitionId);
+      } catch (Exception e) {
+        String errorMessage =
+            String.format("Failed to create snapshot for topic %s partition %d", topicName, partitionId);
+        LOGGER.error(errorMessage, e);
+        throw new VeniceException(errorMessage);
       }
-      concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
+    } else {
       LOGGER.info(
-          "Retrieved snapshot from {} with timestamp {}",
-          fullPathForPartitionDBSnapshot,
-          snapShotTimestamps.get(topicName).get(partitionId));
+          "Snapshot for topic {} partition {} is not stale, skip creating new snapshot. ",
+          topicName,
+          partitionId);
     }
+
+    concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
   }
 
+  /**
+   * Check if the concurrent user count exceeds the limit
+   * @param topicName the topic name
+   * @param partitionId the partition id
+   * @throws VeniceException if the concurrent user count exceeds the limit
+   */
+  private void checkIfConcurrentUserExceedsLimit(String topicName, int partitionId) throws VeniceException {
+    boolean exceededMaxConcurrentUsers = getConcurrentSnapshotUsers(topicName, partitionId) >= MAX_CONCURRENT_USERS;
+    if (exceededMaxConcurrentUsers) {
+      String errorMessage = String.format(
+          "Exceeded the maximum number of concurrent users %d for topic %s partition %d",
+          MAX_CONCURRENT_USERS,
+          topicName,
+          partitionId);
+      LOGGER.error(errorMessage);
+      throw new VeniceException(errorMessage);
+    }
+  }
+
+  /**
+   * Check if the snapshot is stale
+   * @param topicName the topic name
+   * @param partitionId the partition id
+   * @return true if the snapshot is stale, false otherwise
+   */
+  private boolean isSnapshotStale(String topicName, int partitionId) {
+    if (!snapshotTimestamps.containsKey(topicName) || !snapshotTimestamps.get(topicName).containsKey(partitionId)) {
+      return true;
+    }
+    return System.currentTimeMillis()
+        - snapshotTimestamps.get(topicName).get(partitionId) > SNAPSHOT_RETENTION_TIME_IN_MILLIS;
+  }
+
+  /**
+   * Decrease the count of hosts using the snapshot
+   */
   public void decreaseConcurrentUserCount(String topicName, int partitionId) {
```

Comment:
I'm thinking whether we need the lock in
```
    try (AutoCloseableLock ignored = AutoCloseableLock.of(lock)) {
      long result = concurrentUsers.decrementAndGet();
      if (result < 0) {
        throw new VeniceException("Concurrent user count cannot be negative");
      }
```

AtomicLong looks sufficient to me, no?